### PR TITLE
Add /rules/files/{filename} endpoints

### LINKED ISSUE #7178: Add /rules/files/{filename} endpoints
|Wazuh version|Component|
|---|---|
| 4.1.0 | Wazuh API |

Hello team,

as part of the epic issue https://github.com/wazuh/wazuh/issues/7148, we need to split the `manager/files` and `/cluster/{node_id}/files` endpoints.

This issue will cover rule files. For that, we need the following new endpoints:
- [x] `GET /rules/files/{filename}`
  - Add option to choose between XML or JSON.
  - **RBAC**: Action `rules:read` (it already exists). Resource `rule:file`.
- [x] `PUT /rules/files/{filename}`
  - XML format required.
  - **RBAC**: Action `rules:update`. Resource `rule:file`.
- [x] `DELETE /rules/files/{filename}`
  - This endpoint must be restricted to custom rules.
  - **RBAC**: Action `rules:delete`. Resource `rule:file`.

These new endpoints require adaptations in the framework in order to have static paths and proper content validation. Remember to update the documentation for these new RBAC actions and resources.

### wazuh/wazuh
- Unit tests without failures. Updated and/or expanded if there are new functions/methods/outputs:
  - [x] Cluster (`framework/wazuh/core/cluster/tests/` & `framework/wazuh/core/cluster/dapi/tests/`)
  - [x] Core (`framework/wazuh/core/tests/`)
  - [x] SDK (`framework/wazuh/tests/`)
  - [x] RBAC (`framework/wazuh/rbac/tests/`)
  - [x] API (`api/api/tests/`)
- API tavern integration tests without failures. Updated and/or expanded if needed (`api/test/integration/`):
  - [x] Affected tests 
  - [x] Affected RBAC (black and white) tests
- [x] Review of spec.yaml examples and schemas (`api/api/spec/spec.yaml`)
- [x] Changelog (`CHANGELOG.md`)

### wazuh/wazuh-documentation
- [x] Migration from 3.X for changed endpoints (`source/user-manual/api/equivalence.rst`)
- [x] Update RBAC reference with new/modified actions/resources/relationships (`source/user-manual/api/rbac/reference.rst`)

Regards,
Víctor Fernández.

### PR DESCRIPTION
**Pull request by vicferpoy**
Hello team, 
this closes #7178 .

This PR includes the following implementations:
- Refactor of `GET /rules/files/{filename}/download` -> `GET /rules/files/{filename}`
- Development of `PUT` and `DELETE` methods for the endpoint above.
- Improvement of the XML validator function in order to avoid formula injections.
- New unit and integration tests for these endpoints.
- New tools to transform XML files into JSON format.

**NOTE:** this PR adds a new python dependency `xmltodict==0.12.0`. Until we add this dependency to the repository, API integration tests will not pass due to an environment error. If we want to pass them, we need to remove the requirement from `wazuh/framework/requirements.txt` and add the following line to the AIT manager entrypoint before restarting the service:

```
/var/ossec/framework/python/bin/pip3 install xmltodict==0.12.0
``` 

### Unit tests
```
==================================================================================== test session starts ====================================================================================
platform linux -- Python 3.8.5, pytest-5.4.3, py-1.8.2, pluggy-0.13.1
rootdir: /home/vicferpoy/Desktop/Git/wazuh/framework
plugins: metadata-1.10.0, asyncio-0.12.0, cov-2.10.1, html-2.0.1
collected 1609 items                                                                                                                                                                        

core/cluster/dapi/tests/test_dapi.py ......................                                                                                                                           [  1%]
core/cluster/tests/test_cluster.py ...................                                                                                                                                [  2%]
core/cluster/tests/test_common.py .......                                                                                                                                             [  2%]
core/cluster/tests/test_control.py .....                                                                                                                                              [  3%]
core/cluster/tests/test_local_client.py .                                                                                                                                             [  3%]
core/cluster/tests/test_utils.py .......                                                                                                                                              [  3%]
core/cluster/tests/test_worker.py ................                                                                                                                                    [  4%]
core/tests/test_active_response.py ............                                                                                                                                       [  5%]
core/tests/test_agent.py .......................................................................................................................................................      [ 14%]
core/tests/test_cdb_list.py .................................                                                                                                                         [ 16%]
core/tests/test_common.py .......                                                                                                                                                     [ 17%]
core/tests/test_configuration.py ..................................                                                                                                                   [ 19%]
core/tests/test_decoder.py ................                                                                                                                                           [ 20%]
core/tests/test_input_validator.py ...                                                                                                                                                [ 20%]
core/tests/test_logtest.py ..                                                                                                                                                         [ 20%]
core/tests/test_manager.py ................................                                                                                                                           [ 22%]
core/tests/test_ossec_queue.py ...................                                                                                                                                    [ 23%]
core/tests/test_pyDaemonModule.py ......                                                                                                                                              [ 24%]
core/tests/test_results.py ........................................                                                                                                                   [ 26%]
core/tests/test_rootcheck.py ..........                                                                                                                                               [ 27%]
core/tests/test_rule.py ......................                                                                                                                                        [ 28%]
core/tests/test_syscollector.py ...                                                                                                                                                   [ 29%]
core/tests/test_utils.py ............................................................................................................................................................ [ 38%]
......................................................                                                                                                                                [ 42%]
core/tests/test_wazuh_socket.py ....................                                                                                                                                  [ 43%]
core/tests/test_wdb.py .....................                                                                                                                                          [ 44%]
rbac/tests/test_auth_context.py ..                                                                                                                                                    [ 44%]
rbac/tests/test_decorators.py .........................................................................................................                                               [ 51%]
rbac/tests/test_default_configuration.py ....................................................                                                                                         [ 54%]
rbac/tests/test_orm.py .....................................................                                                                                                          [ 57%]
rbac/tests/test_preprocessor.py ...........                                                                                                                                           [ 58%]
tests/test_active_response.py .........                                                                                                                                               [ 59%]
tests/test_agent.py ....................................................................................                                                                              [ 64%]
tests/test_cdb_list.py ...............................................                                                                                                                [ 67%]
tests/test_ciscat.py ..                                                                                                                                                               [ 67%]
tests/test_cluster.py .......                                                                                                                                                         [ 67%]
tests/test_decoders.py ...............................                                                                                                                                [ 69%]
tests/test_group.py ........                                                                                                                                                          [ 70%]
tests/test_logtest.py ......                                                                                                                                                          [ 70%]
tests/test_manager.py ......................................                                                                                                                          [ 72%]
tests/test_mitre.py ................................................................................................................................................................. [ 82%]
...................                                                                                                                                                                   [ 84%]
tests/test_rootcheck.py .................................................                                                                                                             [ 87%]
tests/test_rule.py .......................................................                                                                                                            [ 90%]
tests/test_sca.py .......                                                                                                                                                             [ 90%]
tests/test_security.py .....................................................................                                                                                          [ 95%]
tests/test_stats.py .............                                                                                                                                                     [ 96%]
tests/test_syscheck.py .......................                                                                                                                                        [ 97%]
tests/test_syscollector.py ............                                                                                                                                               [ 98%]
tests/test_task.py ............................                                                                                                                                       [100%]

===================================================================================== warnings summary ======================================================================================
```

### Integration tests
```
test_rbac_black_rule_endpoints
	 8 passed, 10 warnings

test_rbac_white_rule_endpoints
	 8 passed, 10 warnings

test_rule_endpoints
	 15 passed, 17 warnings

```

Regards,
Víctor Fernández

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -55,6 +55,7 @@ All notable changes to this project will be documented in this file.
   - Added debug2 mode for API log and improved debug mode. ([#6822](https://github.com/wazuh/wazuh/pull/6822))
   - Added missing secure headers for API responses. ([#7024](https://github.com/wazuh/wazuh/issues/7024))
   - Added new config option to disable uploading configurations containing remote commands. ([#7016](https://github.com/wazuh/wazuh/issues/7016))
+  - Added new endpoints to manage user rules. ([#7178](https://github.com/wazuh/wazuh/issues/7178))
```
Comment (davidjiglesias) [inline CHANGELOG.md:58]:
```suggestion
  - Added new endpoints to manage rules files. ([#7178](https://github.com/wazuh/wazuh/issues/7178))
```

File: api/api/controllers/rule_controller.py [hunk 2]
```diff
@@ -188,25 +190,115 @@ async def get_rules_files(request, pretty=False, wait_for_complete=False, offset
...
+    Parameters
+    ----------
+    pretty : bool, optional
+        Show results in human-readable format. It only works when `raw` is False (JSON format). Default `True`
+    wait_for_complete : bool, optional
+        Disable response timeout or not. Default `False`
+    filename : str
+        Filename to download.
+    raw : bool, optional
+        Whether to return the file content in raw or JSON format. Default `True`
```
Comment (davidjiglesias) [inline api/api/controllers/rule_controller.py:206]:
```suggestion
        Whether to return the file content in raw or JSON format. Default `False`
```

File: api/api/spec/spec.yaml [hunk 3]
```diff
@@ -365,6 +365,22 @@ x-rbac-catalog:
         actions: ['rules:read']
         resources: ['rule:file:0610-win-ms_logs_rules.xml']
         effect: "allow"
+    'rules:update':
+      description: "Update or upload rule files to Wazuh manager"
+      resources:
+        - $ref: '#/x-rbac-catalog/resources/rule:file'
+      example:
+        actions: ['rules:update']
+        resources: ['rule:file:0610-win-ms_logs_rules.xml']
```
Comment (davidjiglesias) [inline api/api/spec/spec.yaml:374]: I believe the update should be resourceless, can't assign resources to files not yet existent. We can discuss it.

File: api/api/spec/spec.yaml [hunk 4]
```diff
@@ -5182,6 +5198,24 @@ components:
       schema:
         type: string
         format: alphanumeric
+    xml_filename:
+      in: query
+      name: filename
+      description: "Filename of one or more rule or decoder."
```
Comment (davidjiglesias) [inline api/api/spec/spec.yaml:5204]:
```suggestion
      description: "Filter by filename of one or more rule or decoder files."
```

File: api/api/spec/spec.yaml [hunk 5]
```diff
@@ -10913,23 +10947,90 @@ paths:
...
 
-  /rules/files/{filename}/download:
+  /rules/files/{filename}:
     get:
       tags:
         - Rules
-      summary: "Download rule"
-      description: "Download an specified rule file"
-      operationId: api.controllers.rule_controller.get_download_file
+      summary: "Get the content of a rule"
```
Comment (davidjiglesias) [inline api/api/spec/spec.yaml:10954]:
```suggestion
      summary: "Get rules file content"
```

File: api/api/spec/spec.yaml [hunk 6]
```diff
@@ -10995,6 +11096,105 @@ paths:
         '429':
           $ref: '#/components/responses/TooManyRequestsResponse'
 
+    put:
+      tags:
+        - Rules
+      summary: "Update file to the ruleset"
```
Comment (davidjiglesias) [inline api/api/spec/spec.yaml:11102]:
```suggestion
      summary: "Upload rules file"
```

File: api/api/spec/spec.yaml [hunk 7]
```diff
@@ -10995,6 +11096,105 @@ paths:
...
+        '405':
+          $ref: '#/components/responses/InvalidHTTPMethodResponse'
+        '406':
+          $ref: '#/components/responses/WrongContentTypeResponse'
+        '429':
+          $ref: '#/components/responses/TooManyRequestsResponse'
+    delete:
+      tags:
+      - Rules
+      summary: "Delete rule"
```
Comment (davidjiglesias) [inline api/api/spec/spec.yaml:11157]:
```suggestion
      summary: "Delete rules file"
```

**Review by davidjiglesias — CHANGES_REQUESTED** [review]

File: api/api/spec/spec.yaml (hunk 3 above)
Comment (vicferpoy) [reply api/api/spec/spec.yaml:374]: We discussed it and agreed to change it.

**Pull request closed without merging**